compiler/hash-utils/src/range_map.rs

```diff
@@ -0,0 +1,211 @@
+        self.contains(*x)
+    }
+}
+
+impl<T: PrimInt> PartialOrd<T> for Range<T> {
+    fn partial_cmp(&self, x: &T) -> Option<Ordering> {
+        if self.end < *x {
+            Some(Ordering::Less)
+        } else if self.start > *x {
+            Some(Ordering::Greater)
+        } else {
+            Some(Ordering::Equal)
+        }
+    }
+}
+
+impl<I> From<RangeInclusive<I>> for Range<I> {
+    fn from(value: RangeInclusive<I>) -> Self {
+        let (start, end) = value.into_inner();
+        Range { start, end }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RangeMap<I, V> {
+    /// The store that stores the [Range] which maps a range of keys
+    /// to a value.
+    store: Vec<(Range<I>, V)>,
+}
+
+impl<I: fmt::Display, V: fmt::Display> fmt::Display for RangeMap<I, V> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for (key, value) in self.store.iter() {
+            writeln!(f, "{key}: {value}")?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<I: PrimInt + Clone + Copy, V> RangeMap<I, V> {
+    /// Create a new empty [RangeMap].
+    pub fn new() -> Self {
+        Self { store: vec![] }
+    }
+
+    /// Create a new [RangeMap] with a pre-allocated storage of a given
+    /// size.
+    pub fn with_capacity(length: usize) -> Self {
+        Self { store: Vec::with_capacity(length) }
+    }
+
+    /// Create a new [RangeMap] with specified ranges that are assumed
+    /// to be in order.
+    pub fn populated(items: Vec<(RangeInclusive<I>, V)>) -> Self {
+        let map = Self {
+            store: items
+                .into_iter()
+                .map(|(r, v)| (r.into(), v))
+                .sorted_by_key(|(range, _): &(Range<I>, _)| range.start)
+                .collect_vec(),
+        };
+        debug_assert!(map.verify());
+
+        map
+    }
+
+    /// Verify that the [RangeMap] does not contain any overlapping
+    /// ranges.
+    fn verify(&self) -> bool {
+        for pair in self.store.windows(2) {
+            let (left, right) = (pair[0].0, pair[1].0);
+
+            if left.end >= right.start {
+                return false;
+            }
+        }
+
+        true
+    }
+
+    /// Insert a new key-value pair into the [RangeMap].
+    pub fn insert(&mut self, key: RangeInclusive<I>, value: V) {
+        let overlaps = |left: &Range<I>, right: &Range<I>| {
+            if left.start >= right.start && left.end <= right.end {
+                return true;
+            }
+
+            left.contains(right.start) || left.contains(right.end)
+        };
+
+        let key: Range<_> = key.into();
+
+        if self.store.is_empty() {
+            self.store.push((key, value));
+            return;
+        }
+
+        for (item, _) in self.store.iter() {
+            if overlaps(item, &key) {
+                panic!("keys are not allowed to overlap in a range map")
+            }
+
+            if key.start > item.end {
+                continue;
+            }
```

Comment:
why continue here?

Reply:
doesn't this mean that sometimes insert will silently fail?

Reply:
thanks for the catch, fixed!